pkg/runtime/dynamicload.go

```diff
@@ -0,0 +1,102 @@
+package runtime
+
+import (
+	"path/filepath"
+
+	"github.com/dapr/dapr/pkg/components"
+	"github.com/fsnotify/fsnotify"
+
+	componentsV1alpha1 "github.com/dapr/dapr/pkg/apis/components/v1alpha1"
+)
+
+func (a *DaprRuntime) watchPathForDynamicLoading() {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer watcher.Close()
+
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				log.Debug("file event:", event)
+				if event.Op == fsnotify.Create || event.Op == fsnotify.Write {
+					err := a.loadDynamicComponents(event.Name)
+					if err != nil {
+						log.Errorf("failed to load components from file: %s", event.Name)
+					}
+				} else if event.Op == fsnotify.Remove || event.Op == fsnotify.Rename {
+					log.Warnf("manifest file: %s removed from components directory.", filepath.Base(event.Name))
+					for _, comp := range a.dynamicComponents[DynamicComponentsManifest(event.Name)] {
+						log.Warnf("manifest for dynamic loaded component: %s of type: %s removed.", comp.Name, comp.Spec.Type)
+					}
+					log.Warnf("to unload components from dapr, please restart dapr.")
+				}
+			case err, ok := <-watcher.Errors:
+				if !ok {
```

Comment:
ok could be false if it is a zero value generated because the channel is closed and empty. In that case err might not be populated.  https://go.dev/ref/spec#Receive_operator
These are the default max watch numbers given in fsnotify readme https://github.com/fsnotify/fsnotify#platform-specific-notes